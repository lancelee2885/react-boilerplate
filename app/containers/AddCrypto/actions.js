/*
 *
 * AddCrypto actions
 *
 */

import {
  CHANGE_FORM_DATA,
  SUBMIT_FORM_DATA,
  SUBMIT_FORM_DATA_ERROR,
  SUBMIT_FORM_DATA_SUCCESS,
} from './constants';

/**
 * Changes the input fields of the form
 */
export function changeFormData(formData) {
  return {
    type: CHANGE_FORM_DATA,
    formData,
  };
}

/**
 * Changes the input fields of the form
 */
export function submitFormData() {
  return {
    type: SUBMIT_FORM_DATA,
  };
}

/**
 * Changes the input fields of the form
 */
export function formDataSubmitted(formData) {
  return {
    type: SUBMIT_FORM_DATA_SUCCESS,
    formData,
  };
}

/**
 * Changes the input fields of the form
 */
export function formDataError(err) {
  return {
    type: SUBMIT_FORM_DATA_ERROR,
    err,
  };
}
