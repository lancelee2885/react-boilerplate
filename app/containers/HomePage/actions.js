/*
 * App Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

import {
  LOAD_CRYPTO,
  LOAD_CRYPTO_ERROR,
  LOAD_CRYPTO_SUCCESS,
} from './constants';

/**
 * Load the crypto from backend, this action starts the request saga
 *
 * @return {object} An action object with a type of LOAD_CRYPTO
 */
export function loadCryptos() {
  return {
    type: LOAD_CRYPTO,
  };
}

/**
 * Dispatched when the cryptocurrencies are loaded by the request saga
 *
 * @param  {array} cryptos The crypto data
 *
 * @return {object}      An action object with a type of LOAD_CRYPTO_SUCCESS passing the cryptos
 */
export function cryptosLoaded(cryptos) {
  return {
    type: LOAD_CRYPTO_SUCCESS,
    cryptos,
  };
}

/**
 * Dispatched when loading the cryptocurrencies fails
 *
 * @param  {object} error The error
 *
 * @return {object}       An action object with a type of LOAD_CRYPTO_ERROR passing the error
 */
export function cryptossLoadingError(error) {
  return {
    type: LOAD_CRYPTO_ERROR,
    error,
  };
}
