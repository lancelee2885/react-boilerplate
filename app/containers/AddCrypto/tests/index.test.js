/**
 * Test the HomePage
 */

import React from 'react';
import { render } from 'react-testing-library';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';
import { browserHistory } from 'react-router-dom';

import { AddCrypto, mapDispatchToProps } from '../index';
import configureStore from '../../../configureStore';
import { submitFormData, changeFormData } from '../actions';

describe('<AddCrypto />', () => {
  const submittedData = {
    symbol: 'testSymbol',
    name: 'testNamewithsomeextrastringssoitsuniqle',
    description: 'testDescription',
    iconURL: 'testURL',
  };
  let store;

  beforeAll(() => {
    store = configureStore({}, browserHistory);
  });

  it('should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <AddCrypto
            formData={submittedData}
            onChangeFormData={() => null}
            onSubmitForm={() => null}
            isSubmitted
            submitted={submittedData}
            loading
            err={[]}
          />
        </IntlProvider>
      </Provider>,
    );
    expect(firstChild).toMatchSnapshot();
  });

  it('should contain loading spinner if loading', () => {
    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <AddCrypto
            formData={{}}
            onChangeFormData={() => null}
            onSubmitForm={() => null}
            isSubmitted={false}
            submitted={submittedData}
            loading
            err={false}
          />
        </IntlProvider>
      </Provider>,
    );
    expect(firstChild.innerHTML).toContain('Loading');
    expect(firstChild.innerHTML).not.toContain('ErrorMessage');
    expect(firstChild.innerHTML).not.toContain(
      'testNamewithsomeextrastringssoitsuniqle',
    );
  });

  it('should contain error message if there are error(s)', () => {
    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <AddCrypto
            formData={{}}
            onChangeFormData={() => null}
            onSubmitForm={() => null}
            isSubmitted={false}
            submitted={submittedData}
            loading={false}
            err={['error']}
          />
        </IntlProvider>
      </Provider>,
    );
    expect(firstChild.innerHTML).toContain('ErrorMessage');
    expect(firstChild.innerHTML).not.toContain('Loading');
    expect(firstChild.innerHTML).not.toContain(
      'testNamewithsomeextrastringssoitsuniqle',
    );
  });

  it('should contain submitted crypto name if isSubmitted is true', () => {
    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <IntlProvider locale="en">
          <AddCrypto
            formData={{}}
            onChangeFormData={() => null}
            onSubmitForm={() => null}
            isSubmitted
            submitted={submittedData}
            loading={false}
            err={false}
          />
        </IntlProvider>
      </Provider>,
    );
    expect(firstChild.innerHTML).not.toContain('Error Message');
    expect(firstChild.innerHTML).not.toContain('Loading');
    expect(firstChild.innerHTML).toContain(
      'testNamewithsomeextrastringssoitsuniqle',
    );
  });

  describe('mapDispatchToProps', () => {
    describe('onChangeFormData', () => {
      it('should be injected', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        expect(result.onChangeFormData).toBeDefined();
      });

      it('should dispatch changeFormData when called', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        const mockEvt = {
          target: {
            name: 'test',
            value: 'test',
          },
        };
        result.onChangeFormData(mockEvt);
        expect(dispatch).toHaveBeenCalledWith(changeFormData({ test: 'test' }));
      });
    });

    describe('onSubmitForm', () => {
      it('should be injected', () => {
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        expect(result.onSubmitForm).toBeDefined();
      });

      it('should dispatch submitFormData when called', () => {
        const preventDefault = jest.fn();
        const dispatch = jest.fn();
        const result = mapDispatchToProps(dispatch);
        const evt = { preventDefault };
        result.onSubmitForm(evt);
        expect(dispatch).toHaveBeenCalledWith(submitFormData());
      });
    });
  });
});
