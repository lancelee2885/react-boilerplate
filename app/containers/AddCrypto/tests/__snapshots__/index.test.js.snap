// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddCrypto /> should render and match the snapshot 1`] = `
<div
  class="Wrapper__FormWrapper-n4amsx-0 SvHUk"
>
  <div
    class="LoadingElements__Spinner-sc-1f42l8z-0 cJUGse"
  />
  <form
    class="Form-sc-13gsu68-0 ceNKam"
  >
    <label
      for="symbol"
    >
      Symbol:
    </label>
    <input
      class="Input-sc-1o18g32-0 dVYcZi"
      name="symbol"
      required=""
      type="text"
      value="testSymbol"
    />
    <label
      for="name"
    >
      Name:
    </label>
    <input
      class="Input-sc-1o18g32-0 dVYcZi"
      name="name"
      required=""
      type="text"
      value="testNamewithsomeextrastringssoitsuniqle"
    />
    <label
      for="description"
    >
      Description:
    </label>
    <textarea
      class="Input__TextArea-sc-1o18g32-1 gVsiSV"
      name="description"
      required=""
      type="text"
    >
      testDescription
    </textarea>
    <label
      for="icon"
    >
      Icon Link:
    </label>
    <input
      class="Input-sc-1o18g32-0 dVYcZi"
      name="iconURL"
      required=""
      type="url"
      value="testURL"
    />
    <button
      class="SubmitBtnElements__SubmitBtnWrapper-sc-15ad74p-0 dSStNU"
    >
       
      Submit
       
    </button>
  </form>
</div>
`;
